### ISSUE TITLE
css/scss "selector_separator_newline" option doesn't work

### ISSUE THREAD

**Issue by studzinek** [Glavin001/atom-beautify#265]
"selector_separator_newline" is not working with css/scss files  

the same .jsbeautifyrc file with Sublime3 and https://packagecontrol.io/packages/HTML-CSS-JS%20Prettify plugin works as expected

the 'Original File Contents:' after beautify should looks like:
```
body {
 background: red;
}

h1 {
 font-size: 16px;
}
```

I made a mistake or it's a plugin bug?

# Atom Beautify - Debugging information

The following debugging information was generated by `Atom Beautify` on `Wed Apr 01 2015 13:32:51 GMT+0200 (Środkowoeuropejski czas letni)`.

---

**Platform**: win32

## Versions

**Atom Version**: 0.188.0

**Atom Beautify Version**: 0.23.1

## Original file to be beautified

**Original File Path**: `C:\Users\***\Documents\AngularProjects\timesheet-core\app\src\core\directives\menu\test.scss`

**Original File Grammar**: SCSS

**Original File Contents**: 
```SCSS
body {
 background: red;
}
h1 {
 font-size: 16px;
}
```

## Beautification options

**Editor Options**: 
Options from Atom Editor settings
```json
{
    "indent_size": 1,
    "indent_char": "\t",
    "indent_with_tabs": true
}
```

**Config Options**: 
Options from Atom Beautify package settings
```json
{
    "css": {
        "selector_separator_newline": true,
        "indent_size": 1,
        "indent_char": "\t",
        "newline_between_rules": false,
        "preserve_newlines": false
    },
    "js": {
        "indent_size": 1,
        "indent_char": "\t",
        "indent_level": 0,
        "indent_with_tabs": true,
        "preserve_newlines": true,
        "max_preserve_newlines": 10,
        "space_in_paren": false,
        "jslint_happy": false,
        "space_after_anon_function": false,
        "brace_style": "collapse",
        "break_chained_methods": false,
        "keep_array_indentation": false,
        "keep_function_indentation": false,
        "space_before_conditional": true,
        "eval_code": false,
        "unescape_strings": false,
        "wrap_line_length": 0,
        "end_with_newline": false
    },
    "html": {
        "htmlbeautifier_path": "",
        "indent_inner_html": false,
        "indent_size": 1,
        "indent_char": "\t",
        "brace_style": "collapse",
        "indent_scripts": "normal",
        "wrap_line_length": 250,
        "wrap_attributes": "auto",
        "wrap_attributes_indent_size": 1,
        "preserve_newlines": true,
        "max_preserve_newlines": 10,
        "unformatted": [
            "a",
            "sub",
            "sup",
            "b",
            "i",
            "u"
        ],
        "end_with_newline": false
    },
    "sql": {
        "indent_size": 1,
        "keywords": "upper",
        "identifiers": "lower",
        "sqlformat_path": ""
    },
    "markdown": {
        "pandoc_path": "",
        "yaml_front_matter": true
    },
    "perl": {
        "perltidy_path": "perltidy",
        "perltidy_profile": ""
    },
    "php": {
        "cs_fixer_path": "",
        "fixers": "",
        "level": ""
    },
    "python": {
        "autopep8_path": "",
        "max_line_length": 79,
        "indent_size": 1,
        "ignore": [
            "E24"
        ]
    },
    "ruby": {
        "rbeautify_path": ""
    },
    "c": {
        "uncrustifyPath": "",
        "configPath": ""
    },
    "cpp": {
        "uncrustifyPath": "",
        "configPath": ""
    },
    "objectivec": {
        "uncrustifyPath": "",
        "configPath": ""
    },
    "cs": {
        "uncrustifyPath": "",
        "configPath": ""
    },
    "d": {
        "uncrustifyPath": "",
        "configPath": ""
    },
    "java": {
        "uncrustifyPath": "",
        "configPath": ""
    },
    "pawn": {
        "uncrustifyPath": "",
        "configPath": ""
    },
    "vala": {
        "uncrustifyPath": "",
        "configPath": ""
    }
}
```

**Home Options**: 
Options from `C:\Users\***\.jsbeautifyrc`
```json
{}
```

**EditorConfig Options**: 
Options from [EditorConfig](http://editorconfig.org/) file
```json
{}
```

**Project Options**: 
Options from `.jsbeautifyrc` files starting from directory `C:\Users\***\Documents\AngularProjects\timesheet-core\app\src\core\directives\menu` and going up to root
```json
[
    {},
    {},
    {},
    {},
    {},
    {
        "html": {
            "allowed_file_extensions": [
                "htm",
                "html",
                "xhtml",
                "shtml",
                "xml",
                "svg"
            ],
            "brace_style": "collapse",
            "end_with_newline": false,
            "indent_char": "\t",
            "indent_handlebars": false,
            "indent_inner_html": false,
            "indent_scripts": "keep",
            "indent_size": 1,
            "max_preserve_newlines": 0,
            "preserve_newlines": true,
            "unformatted": [
                "a",
                "span",
                "img",
                "code",
                "pre",
                "sub",
                "sup",
                "em",
                "strong",
                "b",
                "i",
                "u",
                "strike",
                "big",
                "small",
                "pre",
                "h1",
                "h2",
                "h3",
                "h4",
                "h5",
                "h6"
            ],
            "wrap_line_length": 120
        },
        "css": {
            "allowed_file_extensions": [
                "css",
                "scss",
                "sass",
                "less"
            ],
            "end_with_newline": false,
            "indent_char": "\t",
            "indent_size": 1,
            "newline_between_rules": true,
            "selector_separator": " ",
            "selector_separator_newline": true
        },
        "js": {
            "allowed_file_extensions": [
                "js",
                "json",
                "jshintrc",
                "jsbeautifyrc"
            ],
            "brace_style": "collapse",
            "break_chained_methods": false,
            "e4x": false,
            "end_with_newline": false,
            "indent_char": "\t",
            "indent_level": 0,
            "indent_size": 1,
            "indent_with_tabs": false,
            "jslint_happy": false,
            "keep_array_indentation": false,
            "keep_function_indentation": false,
            "max_preserve_newlines": 0,
            "preserve_newlines": true,
            "space_after_anon_function": false,
            "space_before_conditional": true,
            "space_in_empty_paren": false,
            "space_in_paren": false,
            "unescape_strings": false,
            "wrap_line_length": 120
        }
    },
    {},
    {},
    {},
    {}
]
```

## Logs

**Error logs**: *Not yet supported*

**Comment by cconover** [comment]
I am seeing a similar issue: if there is a blank line after the closing brace of a declaration, followed by a comment line/block, followed by a new selector, the blank line(s) between the closing brace and the comment are removed. If there is no comment block, the blank line is preserved. If there is a blank  line between the comment and the new selector, the blank line is preserved.

**Comment by Glavin001** [comment]
So CSS and SCSS are handled by two different beautifiers. @Prettydiff handles SCSS and LESS better, so that takes care of those. And CSS is handled by js-beautify, which is what the Sublime3 package uses.

Looking at your sample source code, it looks like plain CSS. I know that the Atom Beautify wrapper of @Prettydiff does not support this `selector_separator_newline`, see https://github.com/Glavin001/atom-beautify/blob/master/lib/langs/css-prettydiff-beautify.coffee however this may be something we can add in the future.

For now, you can beautify in the exact same manner as the Sublime plugin by forcing Atom Beautify to use js-beautify. You can do this by having Atom select the CSS grammar for your file.
On Mac you can do this by pressing `Control + Shift + L` and a list will appear and you can select `CSS`.

**Comment by prettydiff** [comment]
Pretty Diff solves for this with the property `preserve`:

http://prettydiff.com/documentation.php#preserve

The values `css` and `all` would solve this problem. The default value (if the property is completely missing) is `all`, so the new lines should be preserved from the Pretty Diff side. If this does not happen then I have a bug.

**Comment by Glavin001** [comment]
> Pretty Diff solves for this with the property preserve:

Atom Beautify supports Prettydiff's `preserve` option through the option key `preserve_newlines` (see
https://github.com/Glavin001/atom-beautify/blob/master/lib/langs/css-prettydiff-beautify.coffee#L11 ).

> The values css and all would solve this problem. The default value (if the property is completely missing) is all, so the new lines should be preserved from the Pretty Diff side. If this does not happen then I have a bug.

I can confirm that Prettydiff works:

Go to http://prettydiff.com/ and beautify:

```css
body {
 background: red;
}
h1 {
 font-size: 16px;
}
```

and will output correctly:

```css
body {
 background: red;
}
h1 {
 font-size: 16px;
}
```

Atom Beautify turns `preserve` option off if `preserve_newlines` is falsy.

@studzinek : If you can confirm that `preserve_newlines` (`preserve` in Prettydiff) accomplishes what you expect from `selector_separator_newline` then I can make it so that either `preserves_newlines` or `selector_separator_newlines` will also enable `preserve_newlines` in Prettydiff for SCSS and LESS.

**Comment by studzinek** [comment]
preserve_newlines only preserves new lines (between blocks) which I made myself. This option just block removing line but it doesn't add new lines.
Ofcourse preserve_newlines works only with scss files (not css files).


examples and expected behaviour:
With "preserve_newlines": true
before:
```
body {
 background: red;
}

h1 {
 font-size: 16px;
}
```

after beautify:
```
body {
 background: red;
}

h1 {
 font-size: 16px;
}
```

But expected behaviour with selector_separator_newline: true is (with sublime3 it works in this way with scss/css files):
before:
```
body {
 background: red;
}
h1 {
 font-size: 16px;
}
```

after beautify:
```
body {
 background: red;
}

h1 {
 font-size: 16px;
}
```

empty line should be added automatically.

**Comment by prettydiff** [comment]
Adding empty lines where they don't already exist would be an enhancement. https://github.com/prettydiff/prettydiff/issues/72

**Comment by prettydiff** [comment]
Feature added to prettydiff version 1.11.12 with option named `cssinsertlines`

**Comment by Glavin001** [comment]
There are many options that Pretty Diff supports that Atom Beautify does not. This is something I want to tackle head on with #297. See Pretty Diff beautifier class in Atom Beautify: https://github.com/Glavin001/atom-beautify/blob/master/src/beautifiers/prettydiff.coffee

For CSS, js-beautify is used; for SCSS Pretty diff is used. Looks like @prettydiff has already added `cssinsertlines` option. So now it's up to [js-beautify](https://github.com/beautify-web/js-beautify) to have an option to handle this. Atom Beautify passes *all* options directly to js-beautify, meaning that once js-beautify supports a feature and has an option for it, all that needs to be done is update the js-beautify dependency. 
I recommend forwarding this feature request over to https://github.com/beautify-web/js-beautify for CSS. For SCSS, follow #297 and soon we will have complete support for Pretty Diff's options.

In the next release (v0.25.0) you will be able to specify which beautifier is used for each language. This means that since both js-beautify and pretty diff support CSS you could choose.

I am closing this issue in favour of continue in #297 (for SCSS/SASS with Pretty Diff) and waiting on your new issue sent to [js-beautify](https://github.com/beautify-web/js-beautify) for CSS support.

**Comment by Glavin001** [comment]
Just to confirm, this is analogous to JS Beautify's `newline_between_rules`: https://github.com/Glavin001/atom-beautify/blob/master/docs/options.md#css---newline-between-rules

I am resolving #297 now and want to make sure I get them all right. Thanks.

**Comment by Glavin001** [comment]
See available options for `Pretty Diff` beautifier: https://github.com/Glavin001/atom-beautify/blob/master/docs/options.md#pretty-diff

**Comment by Glavin001** [comment]
Published to v0.28.0
